File: SplitLinkedListAlternatingly/SplitLinkedListAlternatingly.cpp
/*
https://www.geeksforgeeks.org/problems/split-singly-linked-list-alternatingly/1
*/

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    struct Node* next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};


// } Driver Code Ends
/*
struct Node
{
    int data;
    struct Node* next;

    Node(int x){
        data = x;
        next = NULL;
    }

};
*/

class Solution {
  public:
    // Function to split a linked list into two lists alternately
    vector<Node*> alternatingSplitList(struct Node* head) {
        vector<int>all;
        while(head!=nullptr){
            all.push_back(head->data);
            head=head->next;
        }
        vector<int>frst, sec;
        int sz=all.size();
        for(int i=0; i<sz; ++i){
            if(i%2==0) frst.push_back(all[i]);
            else sec.push_back(all[i]);
        }
        vector<Node*>res;
        Node *frstHead=new Node(frst[0]);
        Node *tailFrFrst=frstHead;
        for(int i=1; i<frst.size(); ++i){
            Node *nwNode=new Node(frst[i]);
            tailFrFrst->next=nwNode;
            tailFrFrst=tailFrFrst->next;
        }
        Node *secHead=new Node(sec[0]);
        Node *tailFrSec=secHead;
        for(int i=1; i<sec.size(); ++i){
            Node *nwNode=new Node(sec[i]);
            tailFrSec->next=nwNode;
            tailFrSec=tailFrSec->next;
        }
        res.push_back(frstHead);
        res.push_back(secHead);
        return res;
    }
};


//{ Driver Code Starts.

void printList(struct Node* node) {
    while (node != NULL) {
        cout << node->data << " ";
        node = node->next;
    }
    cout << endl;
}

int main() {
    int t;
    cin >> t;
    cin.ignore();

    while (t--) {
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;

        while (ss >> number) {
            arr.push_back(number);
        }

        struct Node* head = new Node(arr[0]);
        struct Node* tail = head;

        for (int i = 1; i < arr.size(); ++i) {
            tail->next = new Node(arr[i]);
            tail = tail->next;
        }

        Solution ob;
        vector<Node*> result = ob.alternatingSplitList(head);
        printList(result[0]);
        printList(result[1]);
    }

    return 0;
}

// } Driver Code Ends

File: SplitLinkedListAlternatingly/SplitLinkedListAlternatingly.java
/*
https://www.geeksforgeeks.org/problems/split-singly-linked-list-alternatingly/1
*/

//{ Driver Code Starts
import java.util.*;

class Node {
    int data;
    Node next;

    Node(int x) {
        data = x;
        next = null;
    }
}

public class Main {
    public static void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        sc.nextLine();

        while (t-- > 0) {
            List<Integer> arr = new ArrayList<>();
            String input = sc.nextLine();
            Scanner ss = new Scanner(input);

            while (ss.hasNextInt()) {
                arr.add(ss.nextInt());
            }

            Node head = new Node(arr.get(0));
            Node tail = head;

            for (int i = 1; i < arr.size(); ++i) {
                tail.next = new Node(arr.get(i));
                tail = tail.next;
            }

            Solution ob = new Solution();
            Node[] result = ob.alternatingSplitList(head);
            printList(result[0]);
            printList(result[1]);
        }

        sc.close();
    }
}

// } Driver Code Ends


// User function Template for Java
/*
struct Node
{
    int data;
    struct Node* next;

    Node(int x){
        data = x;
        next = NULL;
    }

};
*/

class Solution {
    // Function to append a new node with newData at the end of a linked list
    Node[] alternatingSplitList(Node head) {
        ArrayList<Integer>allNodes=new ArrayList<>();
        while(head!=null){
            allNodes.add(head.data);
            head=head.next;
        }
        ArrayList<Integer>firstList=new ArrayList<>();
        ArrayList<Integer>secondList=new ArrayList<>();
        int sizeOfAllNodes=allNodes.size();
        for(int i=0; i<sizeOfAllNodes; ++i){
            if(i%2==0) firstList.add(allNodes.get(i));
            else secondList.add(allNodes.get(i));
        }
        Node[] resultantNodes=new Node[2];
        Node firstHead=new Node(firstList.get(0));
        Node tailForFirst=firstHead;
        for(int i=1; i<firstList.size(); ++i){
            Node newNode=new Node(firstList.get(i));
            tailForFirst.next=newNode;
            tailForFirst=tailForFirst.next;
        }
        Node secondHead=new Node(secondList.get(0));
        Node tailForSecond=secondHead;
        for(int i=1; i<secondList.size(); ++i){
            Node newNode=new Node(secondList.get(i));
            tailForSecond.next=newNode;
            tailForSecond=tailForSecond.next;
        }
        resultantNodes[0]=firstHead;
        resultantNodes[1]=secondHead;
        return resultantNodes;
    }
}